bdai_ros2_wrappers/bdai_ros2_wrappers/subscription.py

```diff
@@ -78,3 +81,137 @@ def wait_for_message(
         future.cancel()
         return None
     return future.result()
+
+
+def wait_for_messages(
+    node: rclpy.node.Node,
+    topics: typing.List,
+    mtypes: typing.List,
+    **kwargs: typing.Any,
+) -> typing.Any:
+    """Waits for messages to arrive at multiple topics within a given time window.
+
+    Uses message_filters.ApproximateTimeSynchronizer. This function blocks
+    until receiving the messages or when a given timeout expires. Assumes the
+    given node is spinning by some external executor.
+
+    Requires the user to pass in a node, since
+    message_filters.Subscriber requires a node upon construction.
+
+    Args:
+        node (Node): the node being attached
+        topics (list): List of topics
+        mtypes (list): List of message types, one for each topic.
+        kwargs: additional arguments, including
+            delay  (float) The delay in seconds for which the messages
+                could be synchronized.
+            allow_headerless (bool): Whether it's ok for there to be
+                no header in the messages.
+            sleep (float) the amount of time to wait before checking
+                whether messages are received
+            timeout (float or None): Time in seconds to wait. None if forever.
+                If exceeded timeout, self.messages will contain None for
+                each topic.
+            latched_topics (set): a set of topics for which the publisher latches (i.e.
+                sets QoS durability to transient_local).
+
+    """
+    return _WaitForMessages(node, topics, mtypes, **kwargs).messages
+
+
+class _WaitForMessages:
+    def __init__(
+        self,
+        node: rclpy.node.Node,
+        topics: typing.List,
+        mtypes: typing.List,
+        queue_size: int = 10,
+        delay: float = 0.2,
+        allow_headerless: bool = False,
+        sleep: float = 0.5,
+        timeout: typing.Optional[float] = None,
+        verbose: bool = False,
+        exception_on_timeout: bool = False,
+        latched_topics: typing.Optional[typing.Set] = None,
+        callback_group: typing.Optional[rclpy.callback_groups.CallbackGroup] = None,
+    ) -> None:
+        self.node = node
+        self.messages: typing.Optional[typing.Tuple] = None
+        self.verbose = verbose
+        self.topics = topics
+        self.timeout = timeout
+        self.exception_on_timeout = exception_on_timeout
+        self.has_timed_out = False
+        if latched_topics is None:
+            latched_topics = set()
+        self.latched_topics = latched_topics
+        self.logger = rclpy.impl.rcutils_logger.RcutilsLogger(name="wait_for_messages")
```

Comment:
@kaiyu-zheng consider using the node logger instead.

Reply:
There is no logger any more, similar to `wait_for_message`. 

It would be nice if `wait_for_future` in `futures.py` can (optionally) log a message periodically, indicating that it's waiting. But this is out of scope for this PR.